fileserver.gr

```diff
@@ -8,59 +8,83 @@ import String from "string"
 import Mediatype from "./lib/mediatype"
 import Stringutil from "./lib/stringutil"
 
-let serve = (abs_path) => {
-    // Trim the leading /
-    let path = String.slice(1, String.length(abs_path), abs_path)
-    File.fdWrite(File.stderr, "Fileserver: Loading file ")
-    File.fdWrite(File.stderr, path)
-    File.fdWrite(File.stderr, "\n")
+let internalError = () => {
+  // Is stdout correct here?
+  let result = File.fdWrite(File.stdout, "Status: 500\n\nInternal Server Error")
+  match (result) {
+    Err(err) => {
+      // TODO: What do you do on a write error?
+      void
+    },
+    Ok(_nBytes) => void,
+  }
+}
 
-    // Open file
-    // TODO: This throws an error when something goes wrong.
-    // But Grain 0.3 does not have a try/catch yet.
-    // When we figure out how, we need to catch the error here
-    // and do a 404 or 500.
-    let input = File.pathOpen(
-        File.pwdfd,
-        [],
-        path,
-        [],
-        [File.FdRead],
-        [],
-        [],   
-    )
+let notFound = () => {
+  let result = File.fdWrite(File.stdout, "Status: 404\n\nNot Found")
+  match (result) {
+    Err(err) => {
+      // TODO: What do you do on a write error?
+      void
+    },
+    Ok(_nBytes) => void,
+  }
+}
 
-    File.fdWrite(File.stdout, "Content-Type: ")
-    File.fdWrite(File.stdout, Mediatype.guess(path))
-    File.fdWrite(File.stdout, "\n\n")
-    
-    // Pipe output to STDOUT
-    let rec pipe = (in, out) => {
-        let (d, len) = File.fdRead(in, 1024)
-        File.fdWrite(out, d)
-        if (len > 0) {
-            pipe(in, out)
-        }
-    }
-    pipe(input, File.stdout)
-    File.fdClose(input)
+// Pipe output to STDOUT
+let rec pipe = (in, out) => {
+  let res = File.fdRead(in, 1024)
+  match (res) {
+    Err(err) => {
+      // TODO: What to do on a read error?
+      void
+    },
+    Ok((d, len)) => {
+      // TODO: What should happen if this Result is an error
+      File.fdWrite(out, d)
+      if (len > 0) {
+        pipe(in, out)
+      }
+    },
+  }
 }
 
-let guestpath = (env) => {
+let serve = abs_path => {
+  // Trim the leading /
+  let path = String.slice(1, String.length(abs_path), abs_path)
+  // TODO: What should happen if any of these Results are an error
+  File.fdWrite(File.stderr, "Fileserver: Loading file ")
+  File.fdWrite(File.stderr, path)
+  File.fdWrite(File.stderr, "\n")
 
-    // Backward compat for an older version of Wagi that had PATH_INFO wrong.
-    // X_RELATIVE_PATH was removed before Wagi 0.4
-    match (Map.get("X_RELATIVE_PATH", env)) {
-        Some(p) => String.concat("/", p),
-        None => {
-            Option.unwrap(Map.get("PATH_INFO", env))
-        }
-    }
-    
+  // Open file
+  let result = File.pathOpen(File.pwdfd, [], path, [], [File.FdRead], [], [])
+
+  match (result) {
+    Err(_err) => notFound(),
+    Ok(input) => {
+      // TODO: What should happen if any of these Results are an error
```

Comment:
Here's what I am thinking -- and I am totally open to alternatives.

- Maybe if the `Content-type` header is concatenated to a string, and then written in one `File.fdWrite` call, we could trap any error and do a `fail`. That way, Wagi will notice this and send the client an HTTP 500.
- If the `pipe` call returns an Err variant but we have already sent the Content-Type header, then I think the best we can do is return. That will result in (at best) partial results or (at worst) an empty body.

As an alternative, if the `pipe` fails, we _can_ still `fail` at this point, because Wagi will not flush the headers until it gets the body (because currently it tries to send Content-Length, and in the future it will try to send the headers and the first chunk, I think). So in that case, even though we have sent the headers to Wagi, if we `fail`, Wagi could still send a 500 to the client.